contracts/cryptoblades.sol

```diff
@@ -451,6 +460,60 @@ contract CryptoBlades is Initializable, AccessControlUpgradeable {
         return (((attacker + 1) % 4) == defender); // Thanks to Tourist
     }
 
+    function _updatePaymentBlockHash() internal {
+        if ((mintPayments[msg.sender].count > 0) &&
+            (mintPayments[msg.sender].blockHash == 0) &&
+            (mintPayments[msg.sender].blockNumber < block.number)) {
+
+            // Payment must be recent enough that the hash is available for the payment block.
+            // Use 200 as a 'friendly' window of "You have 10 minutes."
+            if (mintPayments[msg.sender].blockNumber + 200 >= block.number) {
+                mintPayments[msg.sender].blockHash = blockhash(mintPayments[msg.sender].blockNumber);
+            }
+        }
+    }
+
+    function _discardPaymentIfExpired() internal {
+        _updatePaymentBlockHash();
+        if (mintPayments[msg.sender].count > 0) {
+            if (mintPayments[msg.sender].blockHash == 0) {
+                delete mintPayments[msg.sender];
+            }
+        }
+    }
+
+    function hasPaidForMint() public view returns(bool){
+        return (mintPayments[msg.sender].count > 0);
+    }
+
+    function payForMint(address nftAddress, uint count) public {
+        _discardPaymentIfExpired();
+
+        require(mintPayments[msg.sender].count == 0);
+
+        require(nftAddress == address(weapons));
+        _payContract(msg.sender, mintWeaponFee);
+
+        mintPayments[msg.sender].count = count;
+        mintPayments[msg.sender].nftAddress = nftAddress;
+        mintPayments[msg.sender].blockNumber = block.number;
+        mintPayments[msg.sender].blockHash = 0;
+    }
+
+    function _usePayment(address nftAddress, uint count) internal {
+        _discardPaymentIfExpired();
```

Comment:
I wonder if the discardIfExpired call is necessary if the try portion is to succeed

Reply:
> I wonder if the discardIfExpired call is necessary if the try portion is to succeed

I'm not understanding.  The general idea is that discard actually does two things: (1) discard expire payments and (2) if the payment is present, not expired, and we can update block hash do so.

We could split the discard and make two calls, but that seemed unnecessary.

I'm not clear on your thought here though.  Elaborate?